core/dbt/events/contextvars.py

```diff
@@ -0,0 +1,84 @@
+import contextlib
+import contextvars
```

Comment:
This feels a little bit magical to me — but at least it's pretty standard magic, and not our own hand-rolled magic

https://docs.python.org/3/library/contextvars.html

Reply:
Yeah, there's pretty much no way to do this except by magic :)